rust/main/chains/hyperlane-cosmos/src/providers/cosmos/provider.rs

```diff
@@ -128,23 +132,66 @@ impl CosmosProvider {
 
     /// Extract contract address from transaction.
     /// Assumes that there is only one `MsgExecuteContract` message in the transaction
-    fn contract(tx: &Tx) -> ChainResult<H256> {
+    fn contract(tx: &Tx, tx_hash: &H256) -> ChainResult<H256> {
         use cosmrs::proto::cosmwasm::wasm::v1::MsgExecuteContract as ProtoMsgExecuteContract;
 
-        let any = tx
+        let contract_execution_messages = tx
             .body
             .messages
             .iter()
-            .find(|a| a.type_url == "/cosmwasm.wasm.v1.MsgExecuteContract")
-            .ok_or_else(|| {
-                ChainCommunicationError::from_other_str("could not find contract execution message")
-            })?;
+            .filter(|a| a.type_url == "/cosmwasm.wasm.v1.MsgExecuteContract")
+            .cloned()
+            .collect::<Vec<Any>>();
+
+        if contract_execution_messages.len() > 1 {
+            error!(
+                ?tx_hash,
+                "transaction contains multiple contract execution messages, scraper is using the first entry, manual intervention is required");
+        }
+
+        let any = contract_execution_messages.first().ok_or_else(|| {
+            ChainCommunicationError::from_other_str("could not find contract execution message")
+        })?;
         let proto =
             ProtoMsgExecuteContract::from_any(any).map_err(Into::<HyperlaneCosmosError>::into)?;
         let msg = MsgExecuteContract::try_from(proto)?;
         let contract = H256::try_from(CosmosAccountId::new(&msg.contract))?;
         Ok(contract)
     }
+
+    fn report_unsupported_denominations(&self, tx: &Tx, tx_hash: &H256) {
+        let supported_denomination = self.connection_conf.get_minimum_gas_price().denom;
+        let unsupported_denominations = tx
+            .auth_info
+            .fee
+            .amount
+            .iter()
+            .filter(|c| c.denom.as_ref() != supported_denomination)
+            .map(|c| c.denom.as_ref())
+            .fold("".to_string(), |acc, denom| acc + ", " + denom);
+
+        if !unsupported_denominations.is_empty() {
+            error!(
+                ?tx_hash,
+                ?unsupported_denominations,
+                "transaction contains fees in unsupported denominations, manual intervention is required");
+        }
+    }
+
+    fn convert_fee(&self, coin: &Coin) -> U256 {
+        let gas_price = self.connection_conf.get_minimum_gas_price();
+        if coin.denom.as_ref() != gas_price.denom {
+            return U256::zero();
+        }
+
+        let price = gas_price.amount.parse::<f64>().unwrap_or(1.0);
+        let amount = U256::from(coin.amount);
+        if price < 1.0 {
+            amount * *MICRO_TO_ATTO
```

Comment:
can you provide some more details about what this does?

Reply:
I added a doc comment to the method.

Reply:
hm, so the goal is to scale the fee amount based on the fee denomination of chain we're reading from. It feels more natural to use the prefix in `gas_price.denom` for that, and check it against the [symbol](https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/eb47aaee8627deab2e200bb48ee7df47d2fd33f4/rust/main/config/mainnet_config.json#L2064), assuming we only support native token fees for now. E.g. if the `gasPrice` denom is `untrn` and the symbol is `NTRN`, the `u` prefix should tell us by how much to scale. This would probably make this function easier to understand as well

Reply:
Discussed in person:

1. The problem is that gas price for Neutron and Osmosis is below a single unit of the smallest denomination allowed for the corresponding currency: NTRN and OSMO.
2. Neutron and Osmosis have denomination 6, namely `untrn` and `uosmo` correspondingly for NTRN and OSMO.
3. If we keep gas price in `untrn` and `uosmo`, it will be stored as zero in database since we are using integer values for gas price in database and value after the decimal point will be rounded out.

So, the solution was to scale gas price to `10^-18` so that the gas price is bigger than unity.

Agreed:
1. We continue to scale gas prices for Neutron and Osmosis so that they are expressed in `10^-18` of its corresponding currency. It is inline with gas price in Ethereum which is expressed in wei.
2. We will use `nativeToken` to identify the way to scale instead of relying on the current heuristic: `price < 1.0`.

Reply:
I pushed a commit.